mayastor/src/bdev/nexus/nexus_child.rs

```diff
@@ -252,65 +259,112 @@ impl NexusChild {
             return Err(ChildError::ChildReadOnly {});
         }
 
-        let bdev = self.bdev.as_ref();
-        let desc = self.bdev_handle.as_ref();
-
-        if bdev.is_none() || desc.is_none() {
-            return Err(ChildError::ChildInvalid {});
-        }
+        let bdev = match self.bdev.as_ref() {
+            Some(dev) => dev,
+            None => {
+                return Err(ChildError::ChildInvalid {});
+            }
+        };
 
-        let bdev = bdev.unwrap();
-        let desc = desc.unwrap();
+        let desc = match self.bdev_handle.as_ref() {
+            Some(handle) => handle,
+            None => {
+                return Err(ChildError::ChildInvalid {});
+            }
+        };
 
         let block_size = bdev.block_len();
-
-        let primary = u64::from(block_size);
-        let secondary = bdev.num_blocks() - 1;
-
         let mut buf = desc
             .dma_malloc(block_size as usize)
             .context(LabelAlloc {})?;
 
-        self.read_at(primary, &mut buf)
+        self.read_at(0, &mut buf).await.context(LabelRead {})?;
+        let mbr = match Pmbr::from_slice(&buf.as_slice()[440 .. 512]) {
+            Ok(record) => record,
+            Err(_) => {
+                warn!(
+                    "{}: {}: The protective MBR is invalid!",
+                    self.parent, self.name
+                );
+                return Err(ChildError::LabelInvalid {});
+            }
+        };
+
+        self.read_at(u64::from(block_size), &mut buf)
             .await
             .context(LabelRead {})?;
+        let primary = match GPTHeader::from_slice(buf.as_slice()) {
+            Ok(header) => header,
+            Err(_) => {
+                warn!(
+                    "{}: {}: The primary GPT header is invalid!",
+                    self.parent, self.name
+                );
+                return Err(ChildError::LabelInvalid {});
```

Comment:
If the primary is invalid, why don't we try the second label anymore, as we did before this change?

Reply:
We should check the MBR as well before reconstructing the primary GPT from the secondary.
I had intended to address this fully in CAS-146, but will revert to previous behaviour.

Reply:
Done.
We now recreate the primary from the secondary if possible. We also do the converse as well - if the secondary is not valid then we recreate it from the primary.